File: src/main/java/org/fountainmc/api/world/Location.java
package org.fountainmc.api.world;

import javax.annotation.Nonnull;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

@Nonnull
public final class Location {

    private final World world;
    private final double x, y, z;

    public Location(World world, double x, double y, double z) {
        this.world = checkNotNull(world, "Null world");
        checkArgument(!Double.isInfinite(x), "Infinite x");
        checkArgument(!Double.isInfinite(y), "Infinite y");
        checkArgument(!Double.isInfinite(z), "Infinite z");
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Location add(double x, double y, double z) {
        return new Location(world, this.x + x, this.y + y, this.z + z);
    }

    public double getX() {
        return x;
    }

    public Location withX(double x) {
        return new Location(world, x, y, z);
    }

    public double getY() {
        return y;
    }

    public Location withY(double y) {
        return new Location(world, x, y, z);
    }

    public double getZ() {
        return z;
    }

    public Location withZ(double z) {
        return new Location(world, x, y, z);
    }

    public World getWorld() {
        return world;
    }

    public Location withWorld(World world) {
        return new Location(checkNotNull(world, "Null world"), x, y, z);
    }

    public BlockPosition getBlock() {
        return new BlockPosition(world, (int) Math.floor(x), (int) Math.floor(y), (int) Math.floor(z));
    }

}

File: src/main/kotlin/org/fountainmc/api/world/Location.kt
package org.fountainmc.api.world

import com.google.common.base.Preconditions.checkArgument
import com.google.common.base.Preconditions.checkNotNull

class Location(world: World, val x: Double, val y: Double, val z: Double) {

    val world: World

    init {
        this.world = world
        checkArgument(!java.lang.Double.isInfinite(x), "Infinite x")
        checkArgument(!java.lang.Double.isInfinite(y), "Infinite y")
        checkArgument(!java.lang.Double.isInfinite(z), "Infinite z")
    }

    fun add(x: Double, y: Double, z: Double): Location {
        return Location(world, this.x + x, this.y + y, this.z + z)
    }

    fun withX(x: Double): Location {
        return Location(world, x, y, z)
    }

    fun withY(y: Double): Location {
        return Location(world, x, y, z)
    }

    fun withZ(z: Double): Location {
        return Location(world, x, y, z)
    }

    fun withWorld(world: World): Location {
        return Location(world, x, y, z)
    }

    val block: BlockPosition
        get() = BlockPosition(world, Math.floor(x).toInt(), Math.floor(y).toInt(), Math.floor(z).toInt())

}
